File: api/Enkl.Api/Services/VendorService.cs
using Enkl.Api.Auth;
using Enkl.Api.Data;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Dtos;
using Enkl.Api.Validation;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

/// <summary>
/// OrgAdmin-facing CRUD for Vendor records, plus per-Vendor API key generate/revoke — folded into
/// this one service rather than split out the way OrganisationApiKeyService is its own class (that
/// split exists because the org-wide key is one row shared across many unrelated features; a
/// Vendor's key is intrinsically part of that one Vendor's own record, so keeping it here keeps the
/// surface smaller for a feature this size). Every id is re-validated against the caller's own
/// organisationId before anything is touched — same cross-org-isolation discipline as
/// AnnouncementService/PortfolioCategoryService, no enumeration oracle (a foreign-org Vendor id is
/// silently treated as not-found, never a distinguishable error).
/// </summary>
public class VendorService
{
    private readonly AppDbContext _db;

    public VendorService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<List<VendorDto>> ListAsync(Guid organisationId)
    {
        var vendors = await _db.Vendors.AsNoTracking()
            .Include(v => v.VendorIntegrations)
            .Where(v => v.OrganisationId == organisationId)
            .OrderBy(v => v.Name)
            .ToListAsync();
        return vendors.Select(ToDto).ToList();
    }

    public async Task<VendorDto?> GetAsync(Guid organisationId, Guid vendorId)
    {
        var vendor = await _db.Vendors.AsNoTracking()
            .Include(v => v.VendorIntegrations)
            .FirstOrDefaultAsync(v => v.Id == vendorId && v.OrganisationId == organisationId);
        return vendor is null ? null : ToDto(vendor);
    }

    public async Task<VendorDto> CreateAsync(Guid organisationId, CreateVendorRequest request)
    {
        var name = (request.Name ?? "").Trim();
        if (name.Length == 0) throw new ApiValidationException("Name is required.");

        var now = DateTime.UtcNow;
        var vendor = new Vendor
        {
            Id = Guid.NewGuid(),
            OrganisationId = organisationId,
            Name = name,
            PrimaryContactPerson = Trimmed(request.PrimaryContactPerson),
            ContactEmailAddress = Trimmed(request.ContactEmailAddress),
            ContactUrl = Trimmed(request.ContactUrl),
            TaxNumber = Trimmed(request.TaxNumber),
            DateCreated = now,
            DateLastModified = now
        };
        _db.Vendors.Add(vendor);
        await _db.SaveChangesAsync();
        return ToDto(vendor);
    }

    public async Task<VendorDto?> UpdateAsync(Guid organisationId, Guid vendorId, UpdateVendorRequest request)
    {
        var vendor = await _db.Vendors.Include(v => v.VendorIntegrations)
            .FirstOrDefaultAsync(v => v.Id == vendorId && v.OrganisationId == organisationId);
        if (vendor is null) return null;

        var name = (request.Name ?? "").Trim();
        if (name.Length == 0) throw new ApiValidationException("Name is required.");

        vendor.Name = name;
        vendor.PrimaryContactPerson = Trimmed(request.PrimaryContactPerson);
        vendor.ContactEmailAddress = Trimmed(request.ContactEmailAddress);
        vendor.ContactUrl = Trimmed(request.ContactUrl);
        vendor.TaxNumber = Trimmed(request.TaxNumber);
        vendor.IsActive = request.IsActive;
        vendor.DateLastModified = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        return ToDto(vendor);
    }

    public async Task<bool> DeleteAsync(Guid organisationId, Guid vendorId)
    {
        var vendor = await _db.Vendors.FirstOrDefaultAsync(v => v.Id == vendorId && v.OrganisationId == organisationId);
        if (vendor is null) return false;

        // VendorIntegrations cascade-delete with it (see VendorIntegrationConfiguration) — no
        // separate cleanup needed here.
        _db.Vendors.Remove(vendor);
        await _db.SaveChangesAsync();
        return true;
    }

    /// <summary>Mints a new random key, stores only its hash, and returns the raw value — the one
    /// and only time it's ever retrievable. "Rotate-only": reuses the Vendor's existing
    /// VendorIntegration row if one already exists rather than inserting a second one, so a Vendor
    /// only ever has at most one key — generating a new one immediately invalidates whatever was
    /// issued before, same contract as OrganisationApiKeyService.GenerateAsync, and re-enables the
    /// key if it was previously revoked. Distinct "enklr_vendor_key_" prefix from the org-wide key's
    /// own "enklr_key_" so a leaked key's origin is identifiable at a glance.</summary>
    public async Task<GenerateVendorApiKeyResponse?> GenerateApiKeyAsync(Guid organisationId, Guid vendorId)
    {
        var vendor = await _db.Vendors.Include(v => v.VendorIntegrations)
            .FirstOrDefaultAsync(v => v.Id == vendorId && v.OrganisationId == organisationId);
        if (vendor is null) return null;

        var integration = vendor.VendorIntegrations.FirstOrDefault();
        var now = DateTime.UtcNow;
        if (integration is null)
        {
            integration = new VendorIntegration { Id = Guid.NewGuid(), VendorId = vendor.Id, DateCreated = now };
            _db.VendorIntegrations.Add(integration);
        }

        var rawKey = "enklr_vendor_key_" + Convert.ToBase64String(System.Security.Cryptography.RandomNumberGenerator.GetBytes(32))
            .Replace('+', '-').Replace('/', '_').TrimEnd('=');
        integration.ApiKeyHash = PasswordHasher.Hash(rawKey);
        integration.GeneratedAt = now;
        integration.IsActive = true;
        integration.DateLastModified = now;
        await _db.SaveChangesAsync();

        return new GenerateVendorApiKeyResponse(rawKey);
    }

    public async Task<VendorDto?> RevokeApiKeyAsync(Guid organisationId, Guid vendorId)
    {
        var vendor = await _db.Vendors.Include(v => v.VendorIntegrations)
            .FirstOrDefaultAsync(v => v.Id == vendorId && v.OrganisationId == organisationId);
        if (vendor is null) return null;

        var integration = vendor.VendorIntegrations.FirstOrDefault();
        if (integration is not null)
        {
            // Soft-disable, row kept for audit — same shape as OrganisationApiKeyService.RevokeAsync.
            integration.IsActive = false;
            integration.DateLastModified = DateTime.UtcNow;
            await _db.SaveChangesAsync();
        }
        return ToDto(vendor);
    }

    private static string? Trimmed(string? value)
    {
        if (value is null) return null;
        var trimmed = value.Trim();
        return trimmed.Length == 0 ? null : trimmed;
    }

    private static VendorDto ToDto(Vendor v)
    {
        var integration = v.VendorIntegrations.FirstOrDefault();
        return new VendorDto(
            v.Id, v.Name, v.PrimaryContactPerson, v.ContactEmailAddress, v.ContactUrl, v.TaxNumber,
            v.IsActive, v.DateCreated, v.DateLastModified,
            HasApiKey: !string.IsNullOrEmpty(integration?.ApiKeyHash),
            ApiKeyEnabled: integration?.IsActive ?? false,
            ApiKeyGeneratedAt: integration?.GeneratedAt,
            ApiKeyLastUsedAt: integration?.LastUsedAt);
    }
}

File: php-api/src/Services/VendorService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use Enkl\Api\Auth\PasswordHasher;
use Enkl\Api\Support\Uuid;
use Enkl\Api\Validation\ApiValidationException;
use PDO;

/**
 * Ported from Services/VendorService.cs. OrgAdmin-facing CRUD for Vendor records, plus per-Vendor
 * API key generate/revoke — folded into this one service rather than split out the way
 * OrganisationApiKeyService is its own class (that split exists because the org-wide key is one row
 * shared across many unrelated features; a Vendor's key is intrinsically part of that one Vendor's
 * own record). Every id is re-validated against the caller's own organisationId before anything is
 * touched — same cross-org-isolation discipline as AnnouncementService, no enumeration oracle.
 */
final class VendorService
{
    public function __construct(private readonly PDO $db)
    {
    }

    public function list(string $organisationId): array
    {
        $stmt = $this->db->prepare('SELECT * FROM "Vendors" WHERE "OrganisationId" = :orgId ORDER BY "Name"');
        $stmt->execute(['orgId' => $organisationId]);
        return array_map(fn(array $v) => $this->toDto($v), $stmt->fetchAll());
    }

    public function get(string $organisationId, string $vendorId): ?array
    {
        $vendor = $this->find($vendorId, $organisationId);
        return $vendor === null ? null : $this->toDto($vendor);
    }

    public function create(string $organisationId, array $request): array
    {
        $name = trim((string) ($request['name'] ?? ''));
        if ($name === '') {
            throw new ApiValidationException('Name is required.');
        }

        $id = Uuid::v4();
        $stmt = $this->db->prepare(<<<SQL
            INSERT INTO "Vendors"
                ("Id", "OrganisationId", "Name", "PrimaryContactPerson", "ContactEmailAddress", "ContactUrl", "TaxNumber", "DateCreated", "DateLastModified")
            VALUES (:id, :orgId, :name, :contact, :email, :url, :taxNumber, now(), now())
        SQL);
        $stmt->execute([
            'id' => $id, 'orgId' => $organisationId, 'name' => $name,
            'contact' => $this->trimmedOrNull($request['primaryContactPerson'] ?? null),
            'email' => $this->trimmedOrNull($request['contactEmailAddress'] ?? null),
            'url' => $this->trimmedOrNull($request['contactUrl'] ?? null),
            'taxNumber' => $this->trimmedOrNull($request['taxNumber'] ?? null),
        ]);

        return $this->toDto($this->find($id, $organisationId));
    }

    public function update(string $organisationId, string $vendorId, array $request): ?array
    {
        if ($this->find($vendorId, $organisationId) === null) {
            return null;
        }

        $name = trim((string) ($request['name'] ?? ''));
        if ($name === '') {
            throw new ApiValidationException('Name is required.');
        }

        $stmt = $this->db->prepare(<<<SQL
            UPDATE "Vendors"
            SET "Name" = :name, "PrimaryContactPerson" = :contact, "ContactEmailAddress" = :email,
                "ContactUrl" = :url, "TaxNumber" = :taxNumber, "IsActive" = :isActive, "DateLastModified" = now()
            WHERE "Id" = :id AND "OrganisationId" = :orgId
        SQL);
        $stmt->execute([
            'name' => $name,
            'contact' => $this->trimmedOrNull($request['primaryContactPerson'] ?? null),
            'email' => $this->trimmedOrNull($request['contactEmailAddress'] ?? null),
            'url' => $this->trimmedOrNull($request['contactUrl'] ?? null),
            'taxNumber' => $this->trimmedOrNull($request['taxNumber'] ?? null),
            'isActive' => !empty($request['isActive']) ? 1 : 0,
            'id' => $vendorId, 'orgId' => $organisationId,
        ]);

        return $this->toDto($this->find($vendorId, $organisationId));
    }

    public function delete(string $organisationId, string $vendorId): bool
    {
        // VendorIntegrations cascade-delete with it (ON DELETE CASCADE) — no separate cleanup needed.
        $stmt = $this->db->prepare('DELETE FROM "Vendors" WHERE "Id" = :id AND "OrganisationId" = :orgId');
        $stmt->execute(['id' => $vendorId, 'orgId' => $organisationId]);
        return $stmt->rowCount() > 0;
    }

    /** Mints a new random key, stores only its hash, and returns the raw value — the one and only
     * time it's ever retrievable. "Rotate-only": reuses the Vendor's existing VendorIntegration row
     * if one already exists rather than inserting a second one, so a Vendor only ever has at most
     * one key. Distinct "enklr_vendor_key_" prefix from the org-wide key's own "enklr_key_". */
    public function generateApiKey(string $organisationId, string $vendorId): ?array
    {
        if ($this->find($vendorId, $organisationId) === null) {
            return null;
        }

        $rawKey = 'enklr_vendor_key_' . rtrim(strtr(base64_encode(random_bytes(32)), '+/', '-_'), '=');
        $hash = PasswordHasher::hash($rawKey);

        $integration = $this->findIntegration($vendorId);
        if ($integration !== null) {
            $this->db->prepare(
                'UPDATE "VendorIntegrations" SET "ApiKeyHash" = :hash, "GeneratedAt" = now(), "IsActive" = true, "DateLastModified" = now() WHERE "Id" = :id'
            )->execute(['hash' => $hash, 'id' => $integration['Id']]);
        } else {
            $this->db->prepare(<<<SQL
                INSERT INTO "VendorIntegrations" ("Id", "VendorId", "ApiKeyHash", "GeneratedAt", "IsActive", "DateCreated", "DateLastModified")
                VALUES (:id, :vendorId, :hash, now(), true, now(), now())
            SQL)->execute(['id' => Uuid::v4(), 'vendorId' => $vendorId, 'hash' => $hash]);
        }

        return ['key' => $rawKey];
    }

    public function revokeApiKey(string $organisationId, string $vendorId): ?array
    {
        $vendor = $this->find($vendorId, $organisationId);
        if ($vendor === null) {
            return null;
        }

        $integration = $this->findIntegration($vendorId);
        if ($integration !== null) {
            // Soft-disable, row kept for audit — same shape as OrganisationApiKeyService::revoke.
            $this->db->prepare('UPDATE "VendorIntegrations" SET "IsActive" = false, "DateLastModified" = now() WHERE "Id" = :id')
                ->execute(['id' => $integration['Id']]);
        }

        return $this->toDto($this->find($vendorId, $organisationId));
    }

    private function find(string $vendorId, string $organisationId): ?array
    {
        $stmt = $this->db->prepare('SELECT * FROM "Vendors" WHERE "Id" = :id AND "OrganisationId" = :orgId');
        $stmt->execute(['id' => $vendorId, 'orgId' => $organisationId]);
        $row = $stmt->fetch();
        return $row === false ? null : $row;
    }

    private function findIntegration(string $vendorId): ?array
    {
        $stmt = $this->db->prepare('SELECT * FROM "VendorIntegrations" WHERE "VendorId" = :vendorId LIMIT 1');
        $stmt->execute(['vendorId' => $vendorId]);
        $row = $stmt->fetch();
        return $row === false ? null : $row;
    }

    private function trimmedOrNull(mixed $value): ?string
    {
        if (!is_string($value)) {
            return null;
        }
        $trimmed = trim($value);
        return $trimmed === '' ? null : $trimmed;
    }

    private function toDto(?array $vendor): array
    {
        $integration = $vendor !== null ? $this->findIntegration($vendor['Id']) : null;
        return [
            'id' => $vendor['Id'], 'name' => $vendor['Name'],
            'primaryContactPerson' => $vendor['PrimaryContactPerson'], 'contactEmailAddress' => $vendor['ContactEmailAddress'],
            'contactUrl' => $vendor['ContactUrl'], 'taxNumber' => $vendor['TaxNumber'],
            'isActive' => (bool) $vendor['IsActive'], 'dateCreated' => $vendor['DateCreated'], 'dateLastModified' => $vendor['DateLastModified'],
            'hasApiKey' => $integration !== null && !empty($integration['ApiKeyHash']),
            'apiKeyEnabled' => $integration !== null && (bool) $integration['IsActive'],
            'apiKeyGeneratedAt' => $integration['GeneratedAt'] ?? null,
            'apiKeyLastUsedAt' => $integration['LastUsedAt'] ?? null,
        ];
    }
}
